### DISCUSSION TITLE
Question About using appendChild / other node methods with Alpine JS directives?

### DISCUSSION THREAD

**Discussion by austin-raey** [alpinejs/alpine#3188] [1. Help]
Hey all,

I'm currently using Alpine on a project which requires API calls to retrieve some simple JSON data for components. To add an extra layer of complexity, this data is placed into a Carousel component `x-data="carousel"` and it ultimately shows and hides nodes with native (e.g. `.replaceChildren`) DOM methods.

As an example, I have the following:
```html
<div
  x-data="{
    posts: [],
    init() {
      fetch('https://myapi.com/json-response')
        .then((response) => response.json())
        .then((data) => {
           this.posts = data;
        });
    },
  }"
>
  <template x-if="posts.length">
    <div x-data="carousel">
      <div> ... </div> <!-- Carousel Buttons -->

      <template x-for="(post, i) in posts" :key="i">
        <div class="js-slide" :class="{ hidden: i !== 0 }">
          <i
            x-data="{ icon: {'Facebook': 'fa-facebook', 'Instagram': 'fa-instagram', 'Twitter': 'fa-twitter'}[post.source] }"
            :class="{ 'fa-brands': true, [icon]: true }"
          >
          </i>
          <div x-html="$el.innerHTML || post.message"></div>
        </div>
      </template>
    </div>
  </template>
</div>
```

Strangely, a few things don't work how I expect them to when clicking the Carousel buttons:

- Without `$el.innerHTML || post.message` , all Carousel slides appear to be the first upon navigating. Only the first `post.message` is used.
- The `icon` class appears to be set correctly, however, on the second slide, the class from the first slide is appended for some reason. For example, first slide: `fa-facebook`, second slide `fa-twitter fa-facebook`.

From this behavior, it would seem to me these directives are re-executed when placed into the DOM. Am I missing something to prevent this? This very well could be something I am doing in my Carousel code though.

**Comment by ekwoka** [comment]
Can you show what the carousel component is?

You mention using replace children, but why would you do that?
